[docs] Update limitations page (#19601)
expo/expo: docs/pages/introduction/why-not-expo.mdx

sidebar_title: Limitations
---

import { Collapsible } from '~/ui/components/Collapsible';

> Your success will be limited if you don't know the limitations of your tools. A good software engineer strives to understand the tradeoffs in their decisions.

Reviewer: ```suggestion
We believe Expo tools are a great option for just about anyone looking to create amazing cross-platform apps quickly. Nonetheless, everyone's project is unique, and there may be reasons Expo is not the right choice for what you're building. Read on for some reasons why you may _not_ want to use Expo tools.
```
The original commentary about what is or isn't a "good developer" feels out-of-character with the voice we've been striving for with docs more recently. Just a suggestion above of a possible voicing that's a little more positive.

Author: Thanks for this update! Definitely sounds more positive! 🙏 

I'll just make a small to remove the word "just about" which I'll update on my own.

> ## EAS

Reviewer: ```suggestion
## Expo Application Services
```
It's possible someone can end up here independent of learning anything else about EAS

**With the release of Expo Application Services (EAS)**, building your app with Expo has far fewer limitations, and it can be used for any project, no matter the complexity. With [EAS Build](/build/setup):

<Collapsible summary="All libraries available to React Native apps are available in Expo">

All libraries available to React Native apps are available in Expo projects. Some libraries may require a [Prebuild Config Plugin](/guides/config-plugins/) to be added. The addition of config plugins means:

> - Any third-party push notification service, such as OneSignal, is supported.

Reviewer: A lot of these, but in particular this one, feel like not a limitation at all, but a previous limitation we've addressed. I think the original intent of this doc was to be the "anti-pitch" that's often at the beginning of a tech talk, a way of building trust with the developer by being open about why this thing they're opinionated in favor of may not be right. for them. So, we may want to avoid saying any _specifically_ positive bullet points in order to avoid diluting the anti-pitch. (generalities like "EAS Build is compatible with all RN projects, **but...**" I think are fair game)

- Native libraries that integrate with proprietary services or require configuring native code are supported.

</Collapsible>

<Collapsible summary="App size can be as small as you want">

> Apps built with EAS Build can be as small as you want in size since [the entire Expo SDK is no longer bundled](https://expo.fyi/managed-app-size#app-size-with-eas-build) by default.

Reviewer: A lot of these, but in particular this one, feel like not a limitation at all, but a previous limitation we've addressed. I think the original intent of this doc was to be the "anti-pitch" that's often at the beginning of a tech talk, a way of building trust with the developer by being open about why this thing they're opinionated in favor of may not be right for them. So, we may want to avoid saying any specifically positive bullet points in order to avoid diluting the anti-pitch. (generalities like "EAS Build is compatible with all RN projects, but..." I think are fair game)

Author: I get the idea this is an anti-pitch but we do need a way to convey to the people that the limitations which existed in Expo Go or classic builds aren't true anymore and are resolved with development builds and/or EAS.

Reviewer: I could see still including a sentence or two calling back to limitations we've addressed with new products- something that makes this point, but without most of the EAS Build anti-pitch actually being about Classic Builds. IMO a link to that future blog post idea about a history of the limitations we've addressed would be even better, .e.g, if you want to read how we've evolved things to address previous limitations, check this out.

Author: That sounds good to me. I think after having this discussion with you, I got the fair picture of what we can convert from the Expo Go section. And I totally agree with you that classic build limitation are obsolete and shouldn't be mentioned here.

I'll move the limitations related to classic build system into the archive section.

</Collapsible>

<Collapsible summary="Native dependencies can be customized to target children under a particular age">

Both [Apple](https://developer.apple.com/app-store/review/guidelines/#kids) and [Google](https://support.google.com/googleplay/android-developer/answer/9285070?hl=en) provide strict guidelines for any apps that specifically target children under the age of 13.

Apps built with EAS Build [include only your app's explicit native dependencies](https://blog.expo.dev/expo-managed-workflow-in-2021-d1c9b68aa10). See [Development builds](/build/introduction/) for more information on how to use it.

</Collapsible>

EAS Build is a new service and we plan to address many of the current limitations in time. See [EAS Build Limitations](/build-reference/limitations/) for more information on what are the current limitations.

> ## Expo Go and the classic `expo build` system

Reviewer: I think we should separate out Expo Go and Classic Builds, since the former is sticking around. E.g., "Expo Go is and remains a great way to get started or quickly try something out, but you will need to graduate to dev builds once you use a library with native code that isn't built in."

I'm not even sure we need to mention classic builds in this doc. This is for devs considering Expo, and they shouldn't be jumping into classic builds.

You can have access to dedicated build infrastructure with a ["Priority" plan](https://expo.dev/eas), or you can [run the builds on your own CI](/classic/turtle-cli) if you prefer.

</Collapsible>

<Collapsible summary="Your app cannot target children under 13 years old without customizing native dependencies">

Apps built with `expo build:ios|android` contain code for the entire Expo SDK. Unfortunately, you cannot customize the native dependencies, including Facebook's Audience Network library. So if you build your app this way, you cannot designate it as "designed primarily for children under 13" in the app stores, even though this code is never run unless you explicitly call it.

</Collapsible>

## Bare workflow

> If you'd like to continue getting the same development experience that you have with Expo Go, but also use custom native libraries or custom native code, you should check out the [Development builds]()

> As similar to [Development builds], bare workflow also provides access to the underlying native projects and any native code. It's a "bare" native project where developers make direct changes to their native `android` and `ios` project directories rather than continuously generating them on demand using the [Expo config (**app.json**) and prebuild](/workflow/prebuild/).

Reviewer: Realizing that I may have suggested so many changes that, if you were to accept them all, not much of this section may be left. In general, I think we could summarize bare workflow under EAS Build as something to the effect of "EAS Build still works if you check in your ios and android native project folders, but prebuild can no longer sync changes from your Expo config to the native projects once you do that; you'll have to configure them on your own with native tools like Xcode and Android Studio."

The following list is specifically oriented towards the limitations that exist around using Expo tools and services in the bare workflow.

<Collapsible summary={<span><span className="strike">Build service only works in the managed workflow</span> (Resolved in December 2020)</span>}>

> ~~To build your app binaries for distribution on the Apple App Store and Google Play Store you will need to follow the same steps that you would in any native project, the Expo build service can't handle it for you. We are working on bringing bare workflow support to the build service in the near future.~~

Reviewer: Yeah. It feels like too much detail for something that is deprecated and will no longer be available at all in a few months. I did miss the cross out, and I suppose that would at least prevent it from being cherry-picked.

> You can now use [EAS Build](/build/introduction) to build and sign your apps. [Read the announcement blog post](https://blog.expo.dev/expo-application-services-eas-build-and-submit-fc1d1476aa2e).

Reviewer: ```suggestion
```
I'd be in favor of striking anything that suggests we just changed this yesterday.

</Collapsible>

<Collapsible summary="Configuration must be done on each native project rather than once with Expo config">

Configuring app icons, launch screen, and so on must be configured in the native projects for each platform using the standard native tooling rather than once using a simple JSON object.

</Collapsible>

## Up next

> We are either actively working on or planning to build solutions to all of the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or the [forums](https://forums.expo.dev/).

> If you've been reading along each section of the introduction, you will have a pretty good high-level understanding of Expo tools.

Reviewer: ```suggestion
```

---

import { InlineCode } from '~/components/base/code';
import { BoxLink } from '~/ui/components/BoxLink';

We believe Expo tools are a great option for anyone looking to create amazing cross-platform apps quickly. Nonetheless, everyone's project is unique, and there may be reasons Expo is not the right choice for what you're building. Read on for some reasons why you may _not_ want to use Expo tools.

## EAS

For more information on what current limitations exist with EAS, see the following:

<BoxLink
  title="EAS Build"

>   description="Some of the missing features from EAS could prevent you from being able to use the service for your applications. Learn more."

Reviewer: ```suggestion
  description="EAS Build is designed to work for any React Native project, whether or not you also use Expo Prebuild or manage your own native files. In some situations, you may need to change your project configuration, or it may be incompatible with your app. Learn more."
```
Since this will get quoted, want to make sure we're leading with "it should work with just about anything, but check this stuff to be sure". Also not big on the "missing features" verbiage (we may want to update on the EAS Build limitations page, too), since some of the edge cases in there don't really indicate something that's missing, per se.

  href="/build-reference/limitations/"
/>

### Bare workflow

Similar to [Development builds](/development/introduction), bare workflow also provides access to the underlying native projects and any native code. It's a "bare" native project where you can make direct changes to the project's native **android/** and **ios/** directories rather than continuously generating them on demand using the [Expo config and prebuild](/workflow/prebuild/).

> EAS Build works with bare workflow projects if you check in the native directories but prebuild can no longer sync changes from the Expo config to the native directories. You'll have to configure the native directories on your own with native tools such as Android Studio or Xcode.

Reviewer: ```suggestion
EAS Build is compatible with bare workflow projects if you check in the native directories, but no longer runs prebuild, as that could overwrite any manual customizations you've made to the native project files. You'll have to configure the native directories on your own with native tools such as Android Studio or Xcode.
```

## Expo Go

[Expo Go](/workflow/expo-go) provides a quick way to get started with your app development. It comes with a pre-configured set of libraries known as the Expo SDK. This makes development much faster and makes the mobile development experience much closer to the web development experience.

Like any other tool, it too has its own limitations:

- Using a third-party library that requires native code
- Using a third-party push notification service

> **All of the above have been addressed with EAS Build which is compatible with all type of React Native projects. If you come across them, you can transition your project to use [development builds](/development/introduction/).**

Reviewer: ```suggestion
**We strongly recommend any projects that may need additional libraries with native code migrate to [development builds](/development/introduction/). It's like a version of Expo Go specifically customized to your app's needs.**
```
Or insert verbiage from your new dev client doc :-). At any rate, I think this should reflect that dev builds are for everyone.

## Up next

> We are either actively working on or planning to build solutions to all of the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or the [forums](https://forums.expo.dev/).

Reviewer: ```suggestion
We are either actively working on or planning to build solutions to the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or the [forums](https://forums.expo.dev/).
```
Wanted to couch that statement a little bit, just because some things are limited by their very nature (e.g., Expo Go isn't including custom native code anytime soon)

It might be too much, but wondering if we name-drop Github Discussions as the place for more detailed RfC's/ feature proposals (https://github.com/expo/expo/discussions). Not really committed one way or another on that in this particular spot.

Author: I agree on the replacing forums and canny with GitHub discussions. We do not have a special label or category for forums to accept proposals so I guess directing them to GitHub is better. Would you also like the canny link to be removed?

Reviewer: I think linking to canny and Github for now sounds good. We're still keeping Canny up-to-date at this point

Author: Done!

---

import { InlineCode } from '~/components/base/code';
import { BoxLink } from '~/ui/components/BoxLink';
import { BuildIcon, EasMetadataIcon } from '@expo/styleguide';

> We believe Expo tools are a great option for anyone looking to create amazing cross-platform apps quickly. Nonetheless, everyone's project is unique, and there may be reasons Expo is not the right choice for what you're building. Read on for some reasons why you may **not** want to use Expo tools.

Reviewer: ```suggestion
We believe Expo tools are a great option for anyone looking to create amazing cross-platform apps quickly. Nonetheless, everyone's project is unique, and there may be reasons Expo is not the right choice for what you're building. Read on for some reasons why you may **not** want to use certain Expo tools with your project.
```
Just thinking of those who suggest Expo is a monolith, that you're either in or out, when really you can use just a little or a lot.

EAS Build is compatible with bare workflow projects if you check in the native directories, but no longer runs prebuild, as that could overwrite any manual customizations you've made to the native project files. You'll have to configure the native directories on your own with native tools such as Android Studio or Xcode.

## Expo Go

[Expo Go](/workflow/expo-go) provides a quick way to get started with your app development. It comes with a pre-configured set of libraries known as the Expo SDK. This makes development much faster and makes the mobile development experience much closer to the web development experience.

Like any other tool, it too has its own limitations:

- Using a third-party library that requires native code
- Using a third-party push notification service

**We strongly recommend any projects that require additional libraries with native code to migrate to [development builds](/development/introduction/). It's like creating a version of Expo Go that is specifically customized to your app's needs.**

## Up next

> We are either actively working on or planning to build solutions to all of the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or [GitHub discussions](https://github.com/expo/expo/discussions).

Reviewer: ```suggestion
```
I'm struggling to suggest some alternative wording here. I'm uncomfortable saying these limitations are in our roadmap to resolve, because several of them are either impossible in the current mobile ecosystem (e.g., Expo Go limitations), or involve tradeoffs we're unlikely to accept (e.g., some edge cases in EAS Build).

The wording of "if anything is missing" is tricky, too, because the context above is limitations, but those links are for features. If we're missing a limitation on this page, it probably is actually best to post in forums or edit the doc in a PR, which are features built into the bottom of each page already.

Thus, I was wondering, what if we strike this paragraph altogether?

Author: I agree and too find this wording irrelevant with the current set of tools and features.
